### ISSUE TITLE
add fn read() to Tokenizer

### ISSUE THREAD

**Issue by JohnHMoses** [JohnHMoses/ray_rs#5]
https://pastebin.com/QBep0pau

**Comment by nickswalker** [comment]
I don't think this is possible Rust. As I understand it, the function takes a 'Token' which it uses to check against the type of the next token. If they match, it returns the token, if they don't it returns an error. But there is no way to dynamically check the type of the passed in token. The closest solution, the `Any` trait can't be implemented for non-static lifetimes.

**Comment by nickswalker** [comment]
This is pending support for checking enum discriminants: https://doc.rust-lang.org/std/mem/fn.discriminant.html

**Comment by JohnHMoses** [comment]
I did a preliminary/hacky implementation of `read()` in [this commit](https://github.com/JohnHMoses/ray-rs/commit/e0f5eda0d892d9a654c5dfc4da251c157089ef4b). The only reason this was feasible was because only 3 of our enums actually hold values, so manual, value-agnostic checking was easy to write with pattern matching. Once enum discriminates is stable in rust, we should change our implementation to use that instead.
